src/code/z_vimode.c

```diff
@@ -1,17 +1,329 @@
+    } else if (type == 2) {
+        viMode->customViMode.comRegs.burst = 0x4651E39;
+        viMode->customViMode.comRegs.vSync = 0x20C;
+        viMode->customViMode.comRegs.hSync = 0xC10;
+        viMode->customViMode.comRegs.leap = 0xC1C0C1C;
+        viMode->customViMode.comRegs.hStart = 0x6C02EC;
+        viMode->customViMode.fldRegs[0].vStart = 0x2501FF;
+        viMode->customViMode.fldRegs[0].vBurst = 0xE0204;
+    }
+
+    viMode->customViMode.comRegs.hStart += (unk_left << 16) + (s16)unk_right;
+
+    viMode->customViMode.fldRegs[1].vStart = viMode->customViMode.fldRegs[0].vStart;
+    viMode->customViMode.fldRegs[0].vStart += (unk_top << 16) + (s16)unk_bottom;
+    viMode->customViMode.fldRegs[1].vStart += (unk_top << 16) + (s16)unk_bottom;
+
+    viMode->customViMode.fldRegs[1].vBurst = viMode->customViMode.fldRegs[0].vBurst;
+
+    if (cond_44) {
+        viMode->customViMode.comRegs.vSync++;
+        if (type == 2) {
+            viMode->customViMode.comRegs.hSync += 0x40001;
+        }
+        if (type == 2) {
+            viMode->customViMode.comRegs.leap += 0xFFFCFFFE;
+        }
+    } else {
+        viMode->customViMode.fldRegs[0].vStart += 0xFFFDFFFE;
+        if (type == 2) {
+            viMode->customViMode.fldRegs[0].vBurst += 0xFFFCFFFE;
+        }
+        if (type == 0) {
+            viMode->customViMode.fldRegs[1].vBurst = 0x2FFFE;
+        }
+    }
+
+    viMode->customViMode.comRegs.xScale = (width << 10) / (SCREEN_WIDTH * 2 + unk_right - unk_left);
+    viMode->customViMode.comRegs.vCurrent = 0;
+
+    viMode->customViMode.fldRegs[0].origin = (unk_7C ? 1 : 2) * width * 2;
+    viMode->customViMode.fldRegs[1].origin = (unk_70 ? 1 : 2) * ((unk_7C ? 1 : 2) * width * 2);
+
+    viMode->customViMode.fldRegs[0].yScale = yScaleLo | yScaleHi0;
+    viMode->customViMode.fldRegs[1].yScale = yScaleLo | yScaleHi1;
+
+    viMode->customViMode.fldRegs[0].vIntr = 2;
+    viMode->customViMode.fldRegs[1].vIntr = 2;
+}
+
+#else
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/ViMode_Configure.s")
+#endif
+
+void ViMode_Save(ViMode* viMode) {
+    SREG(48) = viMode->viModeBase;
+    SREG(49) = viMode->viWidth;
+    SREG(50) = viMode->viHeight;
+    SREG(51) = viMode->unk_64;
+    SREG(52) = viMode->unk_60;
+    SREG(53) = viMode->unk_5C;
+    SREG(54) = viMode->unk_58;
+    if (SREG(58) == 1) {
+        SREG(58) = 0;
+        switch (SREG(59)) {
+            case 1:
+                osSyncPrintf("osViModePalLan1\n");
+                ViMode_LogPrint(&osViModePalLan1);
+                break;
+            case 2:
+                osSyncPrintf("osViModeFpalLan1\n");
+                ViMode_LogPrint(&osViModeFpalLan1);
+                break;
+            default:
+                osSyncPrintf("Custom\n");
+                ViMode_LogPrint(&viMode->customViMode);
+                break;
+        }
+    }
+}
+
+void ViMode_Load(ViMode* viMode) {
+    GameInfo* _gameInfo = gGameInfo;
```

Comment:
This seems very fake, is there any way to avoid it, like not using a `width` variable?
Or at least just doing a `if (gGameInfo) {}`?

Reply:
I managed to address all your comments and simplify this function down even more. Thanks!
(It is still a bit weird, because `(x & ~3) == 1` can never be true? it sort of looks like bugged rounding logic?)